### ISSUE TITLE
LoadAsync on Session - when _isNewSessionKey is true

### ISSUE THREAD

**Issue by sitepodmatt** [dotnet/aspnetcore#20898]
https://github.com/dotnet/aspnetcore/blob/8b000d961cd3ccfcc8090fb8368fd6598bace978/src/Middleware/Session/src/DistributedSession.cs

Should Load/LoadAsync be a no-op when _isNewSessionKey is true. Taking the advice to call LoadAsync prior to accessing the session with get/trygetvalue/set  I would of thought we'd guard against unnecsssary chitchat to the cache if we know the session is new?  

I though about doing Request.Cookies["sesscoookie"] != null then await LoadAsync but then doing Get/Set/Id/IsAvailable will simply call load anyway if not _loaded isnt marked.

**Comment by Tratcher** [comment]
That's an interesting point. The case that doesn't cover is if the session database is down, IsAvailable would not be set properly. That might be OK since that's not the general case, or we might have IsAvailable do its own check.

**Comment by sitepodmatt** [comment]
@Tratcher if people are depending on IsAvailable having that do it's own check would be better imo instead of attempting to fetch a known non-existant key for the sake of it.

**Comment by Tratcher** [comment]
IsAvailable would have to do the same call, the distributed cache only offers read and write apis.

**Comment by sitepodmatt** [comment]
I think we're getting our wires crossed here, possibly about different issues? 

What I'm saying is that _cache.Get shouldn't be called if _isNewSessionKey 

```
  var data = _cache.Get(_sessionKey);
                    if (data != null)
                    {
                        Deserialize(new MemoryStream(data));
                    }
                    else if (!_isNewSessionKey)
                    {
                        _logger.AccessingExpiredSession(_sessionKey);
                    }
```

**Comment by Tratcher** [comment]
I understand, it just has side effects on IsAvailable.

**Comment by sitepodmatt** [comment]
_presumeAvailability flag?  I very much doubt people are implementing app/middleware and saying 'before we do work with this freshly basked session let's check the session store is actually available' or branching for both cases

**Comment by Tratcher** [comment]
Yeah, I don't know if anyone is actually using IsAvailable but it was designed to let your site gracefully degrade functionality if the session store was unavailable. E.g. disable the Add-To-Cart button if the cart was going to be stored in session.

**Comment by anurse** [comment]
Flagging this as a `breaking-change`. We *think* it could be done in a non-breaking way but we need to ensure we don't break `IsAvailable`.

**Comment by sitepodmatt** [comment]
Has this been scheduled for .Net 5? I would encourage an alert or warning on the main docs along the lines:

"When optimizing for performance, and thus probably using an async request pipeline, combined with using an external session store be aware the current implementation means that the session store will be "getted" even for newly minted session, the implications (if any) will depend on your setup. " 

I accepts it needs two pre-requistes but in 3.1+ where most developers are favoring an async request pipeline, are those using an external session store such a small number that this was overlooked for years even till today?

**Comment by Tratcher** [comment]
> Has a fix/workaround been scheduled for .Net 5?

No, it's on the backlog.

> where many developers are favoring an async request pipeline

I don't follow, AspNetCore's request pipeline has always been async.

> are those using an external session store such a small number that this is considered a low-priority

It's considered low priority because:
- It only affects performance, not functionality, and the severity of that performance impact has not been demonstrated. I assume fetching a non-existent key from the distributed cache is faster rather than slower than fetching an existing key?
- It's been like this for multiple years and releases without causing significant issues.
- It only affects the first request / new sessions. You'll have to pay the same access cost (or more) on all subsequent requests that do have session data. Amortized out, it's not clear you'd save much in the long run.

I'm marking this as "help wanted" if anyone wants to try fixing it in 5.0.
